File: clang/oneplusone.c
#include "oneplusone.h"
#include "stdio.h"
#include <stdlib.h>

double * mytest;
double ** matrix;

int oneplusoneplus(int val) {
    printf("Hello! one plus one\n");
    printf("your value = %d\n", val);
    return 1 + 1 + val;
}

int show(int *val, int size){
    int sum = 0;
    for (int i=0; i<size; i++) {
        printf("%d\n", val[i]);
        sum = sum + val[i];
    }
    return sum;
}

void print_hello() {
    printf("Hello!!!!! \n");
}

void set_matrix2(int n_rows, int n_cols, double (*mat)[n_cols] ) {
    for (int i=0; i<n_rows; i++) {
        for (int j=0; j<n_cols; j++) {
            printf("%d, %d, %f\n", i, j, mat[i][j]);
        }
    }
}

void set_matrix_from_julia(double ** cmat, double * jmat, int n_rows, int n_cols) {
    for (int i=0; i<n_rows; i++) {
        for (int j=0; j<n_cols; j++) {
            cmat[i][j] = jmat[i + j * n_rows];
        }
    }    
}

void set_matrix(double * mat, int n_rows, int n_cols) {
    printf("n_rows=%d, n_cols=%d \n", n_rows, n_cols);
    
    matrix = malloc(n_rows * sizeof(double *));
    for (int i=0; i<n_rows; i++) {
        matrix[i] = malloc(n_cols * sizeof(double));
    }
    printf("Matrix received from Julia:\n");

    for (int i=0; i<n_rows; i++) {
        for (int j=0; j<n_cols; j++) {
            printf("%3.0f ", mat[i+j*n_rows]);
        }
        printf("\n");
    }    

    set_matrix_from_julia(matrix, mat, n_rows, n_cols);

    printf("Matrix set in C:\n");

    for (int i=0; i<n_rows; i++) {
        for (int j=0; j<n_cols; j++) {
            printf("%3.0f ", matrix[i][j]);
        }
        printf("\n");
    }    
}

double ** get_matrix() {
    printf("getting matrix done!\n");

    return matrix;
}

void set_test(double * val, int len) {
    print_hello();
    mytest = malloc(sizeof(double) * len);
    for (int i=0; i<len; i++) {
        mytest[i] = val[i];
    }
}

double * get_test() {
    print_hello();
    return mytest;
}

File: clang/oneplusone.jl
e = [23, 5, 4, 34, 5, 4]

d = Cint[10, 2, 3, 43, 5]
x = ccall((:show, "liboneplusone"), Cint, (Ptr{Vector{Cint}}, Cint), d, length(d))


len = 1000
vec = collect(100.0:-1:1) .* 10
len = length(vec)

ccall((:set_test, "liboneplusone"), Cvoid, (Ptr{Cdouble}, Cint), vec, len)
y = ccall((:get_test, "liboneplusone"), Ptr{Cdouble}, ())
y_julia = unsafe_wrap(Vector{Cdouble}, y, len; own=false)
@show y_julia

mat = Cdouble[
    1 2 3 4;
    10 20 30 40;
    100 200 300 400
] 


display(mat)
n_rows, n_cols = size(mat)

function Base.cconvert(::Type{Ptr{Ptr{Cdouble}}}, x::Matrix{Cdouble})
    Ref{Ptr{Cdouble}}(
        [Ref(x, i) for i in 1:size(x,1)]
    )
end
# ref_mat = [Ref(mat, i) for i in 1:n_cols:length(mat)]

ccall((:set_matrix, "liboneplusone"), Cvoid, (Ptr{Cdouble}, Cint, Cint), mat, n_rows, n_cols)


mat_c = ccall((:get_matrix, "liboneplusone"), Ptr{Ptr{Cdouble}}, ())
@show typeof(mat_c)

mat_rows = unsafe_wrap(Vector{Ptr{Cdouble}}, mat_c, n_rows)
jrows = Vector{Any}(undef, 0)
for row in mat_rows
    push!(jrows, unsafe_wrap(Vector{Cdouble}, row, n_cols))
end
jmat = zeros(n_rows, n_cols)
for i in 1:n_rows, j in 1:n_cols
    jmat[i,j] = jrows[i][j]
end


